# Segmenting and Clustering Neighborhoods in Toronto

# Part 1: Scraping  Wikipedia page and transforming the data into a pandas dataframe

In [1]:
#Importing necessary libraries

!conda install --yes requests
import requests

!conda install --yes beautifulsoup4
from bs4 import BeautifulSoup

import json 
import numpy as np
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)


!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim

from pandas.io.json import json_normalize

!conda install -c conda-forge matplotlib --yes
import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium 

print('Libraries imported.')



==> WARNING: A newer version of conda exists. <==
  current version: 4.8.3
  latest version: 4.8.4

Please update conda by running

    $ conda update -n base -c defaults conda


ERROR conda.core.link:_execute(700): An error occurred while installing package 'defaults::openssl-1.1.1g-he774522_1'.

[Errno 13] Permission denied: 'C:\\ProgramData\\anaconda3\\envs\\capstone\\Library\\bin\\libssl-1_1-x64.dll'
()



Solving environment: ...working... done

## Package Plan ##

  environment location: C:\ProgramData\anaconda3\envs\capstone

  added / updated specs:
    - requests


The following packages will be UPDATED:

  ca-certificates    conda-forge::ca-certificates-2020.6.2~ --> pkgs/main::ca-certificates-2020.6.24-0

The following packages will be SUPERSEDED by a higher-priority channel:

  certifi            conda-forge::certifi-2020.6.20-py38h3~ --> pkgs/main::certifi-2020.6.20-py38_0
  openssl                                       conda-forge --> pkgs/main


Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
Rolling back transaction: ...working... done
Solving environment: ...working... done

## Package Plan ##

  environment location: C:\ProgramData\anaconda3\envs\capstone

  added / updated specs:
    - beautifulsoup4


The following packages will be UPDATED:

  ca-certificates    conda-forge::ca-certificates-2020



==> WARNING: A newer version of conda exists. <==
  current version: 4.8.3
  latest version: 4.8.4

Please update conda by running

    $ conda update -n base -c defaults conda




==> WARNING: A newer version of conda exists. <==
  current version: 4.8.3
  latest version: 4.8.4

Please update conda by running

    $ conda update -n base -c defaults conda




Solving environment: ...working... done

## Package Plan ##

  environment location: C:\ProgramData\anaconda3\envs\capstone

  added / updated specs:
    - geopy


The following packages will be SUPERSEDED by a higher-priority channel:

  ca-certificates    pkgs/main::ca-certificates-2020.6.24-0 --> conda-forge::ca-certificates-2020.6.20-hecda079_0
  certifi               pkgs/main::certifi-2020.6.20-py38_0 --> conda-forge::certifi-2020.6.20-py38h32f6830_0
  openssl                                         pkgs/main --> conda-forge


Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
Solving environment: ...working... done

# All requested packages already installed.





==> WARNING: A newer version of conda exists. <==
  current version: 4.8.3
  latest version: 4.8.4

Please update conda by running

    $ conda update -n base -c defaults conda




==> WARNING: A newer version of conda exists. <==
  current version: 4.8.3
  latest version: 4.8.4

Please update conda by running

    $ conda update -n base -c defaults conda




Solving environment: ...working... done

# All requested packages already installed.

Libraries imported.


<h3>Reading data from wikipedia url</h3>

In [2]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
data=requests.get(url).text
data

'<!DOCTYPE html>\n<html class="client-nojs" lang="en" dir="ltr">\n<head>\n<meta charset="UTF-8"/>\n<title>List of postal codes of Canada: M - Wikipedia</title>\n<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"3ba1bcbc-be04-44f4-bffc-e0dc62b140d5","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":969510799,"wgRevisionId":969510799,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Short description is different from Wikidata","Communications in Ontar

<h3>Using Beautiful Soup for pulling data out of HTML</h3>

In [3]:
soup = BeautifulSoup(data,'html.parser')
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"3ba1bcbc-be04-44f4-bffc-e0dc62b140d5","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":969510799,"wgRevisionId":969510799,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Short description is different from Wikidata","Communicati

<h3>Getting Table Element from HTML text</h3>

In [4]:
table=soup.find('table')
table

<table class="wikitable sortable">
<tbody><tr>
<th>Postal Code
</th>
<th>Borough
</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park, Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor, Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park, Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue, Humber Valley Village
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern, Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3B
</td>
<td

<h3>Creating Lists of column values from table</h3>

In [5]:
postalCodes=[]
boroughs=[]
neighbourhoods=[]
for rows in table.find_all('tr'):
    columns=rows.find_all('td')
    if len(columns)>0:
        postalCodes.append(columns[0].text.rstrip("\n"))
        boroughs.append(columns[1].text.rstrip("\n"))
        neighbourhoods.append(columns[2].text.rstrip("\n"))

<h3>Creating DataFrame</h3>

In [6]:
df=pd.DataFrame({'Postal Code':postalCodes,'Borough':boroughs,'Neighbourhood':neighbourhoods})
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

<h3>Removing records having value "Not assigned" value in Borough Column</h3>

In [7]:
df.drop(df[df['Borough']=='Not assigned'].index,inplace=True)
df.reset_index(drop=True,inplace=True)
df.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

<h3>Checking no of records having "Not assigned" in Neighbourhood Column<h3>

In [8]:
df[df['Neighbourhood'] == 'Not assigned'].shape[0]

0

Here we see that there is no such record having ```Not assigned``` in Neighbourhood Column. So we don't need to do any transformation

<h3>Combining rows having same Postal Code and Borough into single row</h3>

In [9]:
df=df.groupby(['Postal Code','Borough'])['Neighbourhood'].apply(','.join).reset_index()
df.head()

Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

<h3>Validating record with the screenshot given in the Question</h3>

In [10]:
test_df=pd.DataFrame(columns=df.columns)

test_list = ["M5G", "M2H", "M4B", "M1J", "M4G", "M4M", "M1R", "M9V", "M9L", "M5V", "M1B", "M5A"]

for postcode in test_list:
    test_df = test_df.append(df[df["Postal Code"]==postcode], ignore_index=True)
    
test_df

Postal Code           Borough  \
0          M5G  Downtown Toronto   
1          M2H        North York   
2          M4B         East York   
3          M1J       Scarborough   
4          M4G         East York   
5          M4M      East Toronto   
6          M1R       Scarborough   
7          M9V         Etobicoke   
8          M9L        North York   
9          M5V  Downtown Toronto   
10         M1B       Scarborough   
11         M5A  Downtown Toronto   

                                        Neighbourhood  
0                                  Central Bay Street  
1                                   Hillcrest Village  
2                     Parkview Hill, Woodbine Gardens  
3                                 Scarborough Village  
4                                             Leaside  
5                                     Studio District  
6                                   Wexford, Maryvale  
7   South Steeles, Silverstone, Humbergate, Jamest...  
8                                       Humber Summit  
9   CN Tower, King and Spadina, Railway Lands, Har...  
10                                     Malvern, Rouge  
11                          Regent Park, Harbourfront

And we found exactly same.

<h3>Number of rows in dataframe</h3>

In [11]:
df.shape

(103, 3)

Here we see that there are 103 rows in our dataframe.

# Part 2: Finding  latitude and the longitude coordinates of each neighborhood

<h3>Reading Record from CSV</h3>

In [12]:
newdf=pd.read_csv('Geospatial_Coordinates.csv')
newdf.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

<h3>Creating Final Dataframe with Latitutude and Longitudes of each Neighbourhood</h3>

In [13]:
latitudes=[]
longitudes=[]
for postal_code in df['Postal Code']:
    lat_long=newdf[newdf['Postal Code']==postal_code]    
    latitudes.append(lat_long['Latitude'].tolist()[0])
    longitudes.append(lat_long['Longitude'].tolist()[0])

df_lg=pd.DataFrame({'Latitude':latitudes,'Longitude':longitudes})
df_lat_long=pd.concat([df,df_lg],axis=1)
df_lat_long.head()

Postal Code      Borough                           Neighbourhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

<h3>Validating record with the screenshot given in the Question</h3>

In [14]:
new_test_df=pd.DataFrame(columns=df_lat_long.columns)

test_list = ["M5G", "M2H", "M4B", "M1J", "M4G", "M4M", "M1R", "M9V", "M9L", "M5V", "M1B", "M5A"]
for postcode in test_list:
    new_test_df = new_test_df.append(df_lat_long[df_lat_long["Postal Code"]==postcode], ignore_index=True)
    
new_test_df

Postal Code           Borough  \
0          M5G  Downtown Toronto   
1          M2H        North York   
2          M4B         East York   
3          M1J       Scarborough   
4          M4G         East York   
5          M4M      East Toronto   
6          M1R       Scarborough   
7          M9V         Etobicoke   
8          M9L        North York   
9          M5V  Downtown Toronto   
10         M1B       Scarborough   
11         M5A  Downtown Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                                  Central Bay Street  43.657952 -79.387383  
1                                   Hillcrest Village  43.803762 -79.363452  
2                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
3                                 Scarborough Village  43.744734 -79.239476  
4                                             Leaside  43.709060 -79.363452  
5                                     Studio District  43.659526 -79.340923  
6                                   Wexford, Maryvale  43.750072 -79.295849  
7   South Steeles, Silverstone, Humbergate, Jamest...  43.739416 -79.588437  
8                                       Humber Summit  43.756303 -79.565963  
9   CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
10                                     Malvern, Rouge  43.806686 -79.194353  
11                          Regent Park, Harbourfront  43.654260 -79.360636

# Part3: Explore and cluster the neighborhoods in Toronto

#### Use geopy library to get the latitude and longitude values of  Toronto.

In [15]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toranto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toranto are 43.6534817, -79.3839347.


#### Create a map of Toranto with neighbourhoods superimposed on it

In [16]:
map_Toranto=folium.Map(location=[latitude,longitude],zoom_start=10)

for lat,long,borough,neighbourhood in zip(df_lat_long['Latitude'],df_lat_long['Longitude'],df_lat_long['Borough'],df_lat_long['Neighbourhood']):
    label='{},{}'.format(neighbourhood,borough)
    label=folium.Popup(label,parse_html=True)
    folium.CircleMarker(
    [lat,long],
    radius=5,
    popup=label,
    color='blue',
    fill=True,
    fill_color='#3186cc',
    fill_opacity=0.7,
    parse_html=False).add_to(map_Toranto)
    
map_Toranto

#### Exploring Neighbourhoods and segment and cluter in Downtown Toranto

In [17]:
downtown_data = df_lat_long[df_lat_long['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
downtown_data.head()

Postal Code           Borough                Neighbourhood   Latitude  \
0         M4W  Downtown Toronto                     Rosedale  43.679563   
1         M4X  Downtown Toronto  St. James Town, Cabbagetown  43.667967   
2         M4Y  Downtown Toronto         Church and Wellesley  43.665860   
3         M5A  Downtown Toronto    Regent Park, Harbourfront  43.654260   
4         M5B  Downtown Toronto     Garden District, Ryerson  43.657162   

   Longitude  
0 -79.377529  
1 -79.367675  
2 -79.383160  
3 -79.360636  
4 -79.378937

#### Getting Longitude and Latitude of Downtown Toranto

In [18]:
address = 'Downtown Toronto'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toranto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toranto are 43.6541737, -79.38081164513409.


#### Creating map of Downtown Toranto

In [19]:
# create map of Downtown using latitude and longitude values
map_downtown = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(downtown_data['Latitude'], downtown_data['Longitude'], downtown_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown)  
    
map_downtown

#### Foursuare Credentials and Version

In [20]:
CLIENT_ID = 'XB211O2DGNOE20ZFWAC1MT4U1HZNOLRJ4L3LL1Y03NAISJRZ' 
CLIENT_SECRET = '2UOVL44N5AKKEVNXJAB0A01EITNSZWWGC1CZUJ051CK2HNHT' 
VERSION = '20200814' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: XB211O2DGNOE20ZFWAC1MT4U1HZNOLRJ4L3LL1Y03NAISJRZ
CLIENT_SECRET:2UOVL44N5AKKEVNXJAB0A01EITNSZWWGC1CZUJ051CK2HNHT


#### Explore First Neighbourhood in  Downtown Toranto

In [21]:
downtown_data.loc[0, 'Neighbourhood']

'Rosedale'

In [22]:
neighborhood_latitude = downtown_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = downtown_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = downtown_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Rosedale are 43.6795626, -79.37752940000001.


In [23]:
##Getting top 100 Venues that are in  first neighbourhood(i.e. Rosedale) within radius=500
latitude=neighborhood_latitude
longitude=neighborhood_longitude
radius=500
LIMIT=100
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=XB211O2DGNOE20ZFWAC1MT4U1HZNOLRJ4L3LL1Y03NAISJRZ&client_secret=2UOVL44N5AKKEVNXJAB0A01EITNSZWWGC1CZUJ051CK2HNHT&ll=43.6795626,-79.37752940000001&v=20200814&radius=500&limit=100'

In [24]:
#Send Get Request and examine result
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5f38a9ebcf2d854abd251ebe'},
 'response': {'headerLocation': 'Rosedale',
  'headerFullLocation': 'Rosedale, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.6840626045, 'lng': -79.37131878274371},
   'sw': {'lat': 43.675062595499995, 'lng': -79.38374001725632}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4aff2d47f964a520743522e3',
       'name': 'Rosedale Park',
       'location': {'address': '38 Scholfield Ave.',
        'crossStreet': 'at Edgar Ave.',
        'lat': 43.68232820227814,
        'lng': -79.37893434347683,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.68232820227814,
          'lng': -79.37893434347683}],
        'distance': 32

In [25]:
#Method to get category name
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [26]:
#Transforming Json to Pandas DataFrame
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

<ipython-input-26-ed183576ee8b>:4: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  nearby_venues = json_normalize(venues) # flatten JSON


name  categories        lat        lng
0         Rosedale Park  Playground  43.682328 -79.378934
1          Whitney Park        Park  43.682036 -79.373788
2  Alex Murray Parkette        Park  43.678300 -79.382773
3        Milkman's Lane       Trail  43.676352 -79.373842

In [27]:
#Total venues in Rosedale
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

4 venues were returned by Foursquare.


#### Exploring all neighbourhoods in DownTown Toronto

In [28]:
#Methods to get all venues in nearby neighbourhoods
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *downtown_venues*.

In [29]:
downtown_venues=getNearbyVenues(downtown_data["Neighbourhood"],downtown_data["Latitude"],downtown_data["Longitude"])

Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Queen's Park, Ontario Provincial Government


#### Checking size of new dataframe

In [30]:
print(downtown_venues.shape)
downtown_venues.head()

(1237, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                     Rosedale              43.679563              -79.377529   
1                     Rosedale              43.679563              -79.377529   
2                     Rosedale              43.679563              -79.377529   
3                     Rosedale              43.679563              -79.377529   
4  St. James Town, Cabbagetown              43.667967              -79.367675   

                  Venue  Venue Latitude  Venue Longitude Venue Category  
0         Rosedale Park       43.682328       -79.378934     Playground  
1          Whitney Park       43.682036       -79.373788           Park  
2  Alex Murray Parkette       43.678300       -79.382773           Park  
3        Milkman's Lane       43.676352       -79.373842          Trail  
4           Cranberries       43.667843       -79.369407          Diner

#### venues counts for each neighbourhood

In [31]:
downtown_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     65   
Christie                                                               17   
Church and Wellesley                                                   74   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              64   
Queen's Park, Ontario Provincial Government                            36   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
St. James Town                                                         85   
St. James Town, Cabbagetown                                            42   
Stn A PO Boxes                                                         96   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         36   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      65   
Christie                                                                17   
Church and Wellesley                                                    74   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               64   
Queen's Park, Ontario Provincial Government                             36   
Regent Park, Harbourfront                                               45   
Richmond, Adelaide, King                                               100   
Rosedale                                                                 4   
St. James Town                                                          85   
St. James Town, Cabbagetown                                             42   
Stn A PO Boxes                                                          96   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          36   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            58              58   
CN Tower, King and Spadina, Railway Lands, Harb...     15              15   
Central Bay Street                                     65              65   
Christie                                               17              17   
Church and Wellesley                                   74              74   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

#### Let's find out how many unique categories can be curated from all the returned venues

In [32]:
print('There are {} uniques categories.'.format(len(downtown_venues['Venue Category'].unique())))

There are 213 uniques categories.


### Analyze each Neighbourhood

In [33]:
# one hot encoding
downtown_onehot = pd.get_dummies(downtown_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
downtown_onehot['Neighbourhood'] = downtown_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [downtown_onehot.columns[-1]] + list(downtown_onehot.columns[:-1])
downtown_onehot =downtown_onehot[fixed_columns]

downtown_onehot.head()

Neighbourhood  Afghan Restaurant  Airport  \
0                     Rosedale                  0        0   
1                     Rosedale                  0        0   
2                     Rosedale                  0        0   
3                     Rosedale                  0        0   
4  St. James Town, Cabbagetown                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bike Rental / Bike Share  Bistro  Boat or Ferry  \
0                   0                         0       0              0   
1                   0                         0       0              0   
2                   0                         0       0              0   
3                   0                         0       0              0   
4                   0                         0       0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Butcher  Café  \
0                0         0             0              0        0     0   
1                0         0             0              0        0     0   
2                0         0             0              0        0     0   
3                0         0             0              0        0     0   
4                0         0             0              0        0     0   

   Camera Store  Candy Store  Caribb

#### Checking size of new dataframe

In [34]:
downtown_onehot.shape

(1237, 214)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [35]:
downtown_grouped = downtown_onehot.groupby('Neighbourhood').mean().reset_index()
downtown_grouped

Neighbourhood  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
2                                  Central Bay Street           0.000000   
3                                            Christie           0.000000   
4                                Church and Wellesley           0.013514   
5                      Commerce Court, Victoria Hotel           0.000000   
6              First Canadian Place, Underground city           0.000000   
7                            Garden District, Ryerson           0.000000   
8   Harbourfront East, Union Station, Toronto Islands           0.000000   
9           Kensington Market, Chinatown, Grange Park           0.000000   
10        Queen's Park, Ontario Provincial Government           0.000000   
11                          Regent Park, Harbourfront           0.000000   
12                           Richmond, Adelaide, King           0.000000   
13                                           Rosedale           0.000000   
14                                     St. James Town           0.000000   
15                        St. James Town, Cabbagetown           0.000000   
16                                     Stn A PO Boxes           0.000000   
17           Toronto Dominion Centre, Design Exchange           0.000000   
18                     University of Toronto, Harbord           0.000000   

     Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   0.000000            0.000000      0.000000        0.000000   
1   0.066667            0.066667      0.066667        0.133333   
2   0.000000            0.000000      0.000000        0.000000   
3   0.000000            0.000000      0.000000        0.000000   
4   0.000000            0.000000      0.000000        0.000000   
5   0.000000            0.000000      0.000000        0.000000   
6   0.000000            0.000000      0.000000        0.000000   
7   0.000000            0.000000      0.000000        0.000000   
8   0.000000            0.000000      0.000000        0.000000   
9   0.000000            0.000000      0.000000        0.000000   
10  0.000000            0.000000      0.000000        0.000000   
11  0.000000            0.000000      0.000000        0.000000   
12  0.000000            0.000000      0.000000        0.000000   
13  0.000000            0.000000      0.000000        0.000000   
14  0.000000            0.000000      0.000000        0.000000   
15  0.000000            0.000000      0.000000        0.000000   
16  0.000000            0.000000      0.000000        0.000000   
17  0.000000            0.000000      0.000000        0.000000   
18  0.000000            0.000000      0.000000        0.000000   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0               0.0          0.000000             0.000000      0.000000   
1               0.2          0.066667             0.000000      0.000000   
2               0.0          0.000000             0.000000      0.000000   
3               0.0          0.000000             0.000000      0.000000   
4               0.0          0.000000             0.013514      0.000000   
5               0.0          0.000000             0.040000      0.000000   
6               0.0          0.000000             0.030000      0.000000   
7               0.0          0.000000             0.000000      0.000000   
8               0.0          0.000000             0.000000      0.000000   
9               0.0          0.000000             0.000000      0.000000   
10              0.0          0.000000             0.000000      0.000000   
11              0.0          0.000000             0.000000      0.022222   
12              0.0          0.000000             0.020000      0.000000   
13              0.0          0.000000             0.000000      0.000000   
14              0.0          0.000000             0.035294

#### Confirm new size

In [36]:
downtown_grouped.shape

(19, 214)

#### Getting top 5 venues of each neighbourhood

In [37]:
num_top_venues=5

for hood in downtown_grouped["Neighbourhood"]:
    print("----"+hood+"----")
    temp=downtown_grouped[downtown_grouped["Neighbourhood"]==hood].T.reset_index()
    temp=temp.iloc[1:]
    temp.columns = ['venue','freq']
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
         venue  freq
0  Coffee Shop  0.09
1         Café  0.03
2     Beer Bar  0.03
3       Bakery  0.03
4   Restaurant  0.03


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                 venue  freq
0      Airport Service  0.20
1       Airport Lounge  0.13
2      Harbor / Marina  0.07
3              Airport  0.07
4  Rental Car Location  0.07


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.17
1                Café  0.06
2      Sandwich Place  0.05
3  Italian Restaurant  0.05
4    Department Store  0.03


----Christie----
           venue  freq
0  Grocery Store  0.24
1           Café  0.18
2           Park  0.12
3     Restaurant  0.06
4    Candy Store  0.06


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.11
1  Japanese Restaurant  0.07
2              Gay Bar  0.05
3     Sushi Restaurant  0.05
4           Restaurant  0.04


--

#### Putting top 10 venues of each neighbourhood in DataFrame

In [38]:
#Method to sort the venues category in descending order
def return_most_common_venues(row, num_top_venues):    
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Creating dataframe and displaying 10 venues for each neighbourhood

In [39]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = downtown_grouped['Neighbourhood']

for ind in np.arange(downtown_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Farmers Market                Bakery           Cheese Shop   
1        Airport Lounge       Harbor / Marina                 Plane   
2                  Café        Sandwich Place    Italian Restaurant   
3                  Café                  Park             Nightclub   
4   Japanese Restaurant               Gay Bar      Sushi Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Restaurant    Seafood Restaurant              Beer Bar   
1         Boat or Ferry   Rental Car Location              Boutique   
2      Department Store           Salad Place          Burger Joint   
3                 Diner            Restaurant    Italian Restaurant   
4            Restaurant                   Pub           Men's Store   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Cocktail Bar                  Café       Greek Restaurant  
1          Sculpture Garden               Airport       Airport Terminal  
2           Bubble Tea Shop   Japanese Restaurant            Yoga Studio  
3                Baby Store           Candy Store     Athletics & Sports  
4  Mediterranean Restaurant                 Hotel            Yoga Studio


### Cluster Neighbourhoods

In [40]:
#Run k-means to cluster the neighborhood into 5 clusters.

# set number of clusters
kclusters = 5

downtown_grouped_clustering = downtown_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(downtown_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 3, 2, 4, 2, 2, 2, 2, 2, 2])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [41]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

downtown_merged = downtown_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
downtown_merged = downtown_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

downtown_merged.head() # check the last columns!

Postal Code           Borough                Neighbourhood   Latitude  \
0         M4W  Downtown Toronto                     Rosedale  43.679563   
1         M4X  Downtown Toronto  St. James Town, Cabbagetown  43.667967   
2         M4Y  Downtown Toronto         Church and Wellesley  43.665860   
3         M5A  Downtown Toronto    Regent Park, Harbourfront  43.654260   
4         M5B  Downtown Toronto     Garden District, Ryerson  43.657162   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.377529               1                  Park            Playground   
1 -79.367675               2           Coffee Shop           Pizza Place   
2 -79.383160               2           Coffee Shop   Japanese Restaurant   
3 -79.360636               0           Coffee Shop                Bakery   
4 -79.378937               2        Clothing Store           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Trail           Yoga Studio          Dance Studio   
1                Bakery                   Pub                Market   
2               Gay Bar      Sushi Restaurant            Restaurant   
3                   Pub                  Café                  Park   
4                  Café       Bubble Tea Shop   Japanese Restaurant   

  6th Most Common Venue 7th Most Common Venue     8th Most Common Venue  \
0   Dumpling Restaurant            Donut Shop          Doner Restaurant   
1    Italian Restaurant                  Café                Restaurant   
2                   Pub           Men's Store  Mediterranean Restaurant   
3        Breakfast Spot               Theater               Yoga Studio   
4        Cosmetics Shop    Italian Restaurant                     Plaza   

  9th Most Common Venue 10th Most Common Venue  
0               Dog Run    Distribution Center  
1   American Restaurant               Pharmacy  
2                 Hotel            Yoga Studio  
3        Cosmetics Shop             Shoe Store  
4           Pizza Place              Bookstore

#### Visualizing Clusters

In [42]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(downtown_merged['Latitude'], downtown_merged['Longitude'], downtown_merged['Neighbourhood'], downtown_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine Clusters

In [43]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 0, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
3   Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
3                 Bakery                   Pub                  Café   
18           Yoga Studio    Mexican Restaurant         Smoothie Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
3                   Park        Breakfast Spot               Theater   
18    Italian Restaurant              Beer Bar   Japanese Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
3            Yoga Studio        Cosmetics Shop             Shoe Store  
18        Sandwich Place   Distribution Center                  Diner

In [44]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 1, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0  Downtown Toronto               1                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Playground                 Trail           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Dance Studio   Dumpling Restaurant            Donut Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant               Dog Run    Distribution Center

In [45]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 2, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1   Downtown Toronto               2           Coffee Shop   
2   Downtown Toronto               2           Coffee Shop   
4   Downtown Toronto               2        Clothing Store   
5   Downtown Toronto               2                  Café   
6   Downtown Toronto               2           Coffee Shop   
7   Downtown Toronto               2           Coffee Shop   
8   Downtown Toronto               2           Coffee Shop   
9   Downtown Toronto               2           Coffee Shop   
10  Downtown Toronto               2           Coffee Shop   
11  Downtown Toronto               2           Coffee Shop   
12  Downtown Toronto               2                  Café   
13  Downtown Toronto               2           Coffee Shop   
15  Downtown Toronto               2           Coffee Shop   
16  Downtown Toronto               2           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1            Pizza Place                Bakery                    Pub   
2    Japanese Restaurant               Gay Bar       Sushi Restaurant   
4            Coffee Shop                  Café        Bubble Tea Shop   
5            Coffee Shop        Clothing Store         Cosmetics Shop   
6         Farmers Market                Bakery            Cheese Shop   
7                   Café        Sandwich Place     Italian Restaurant   
8                   Café            Restaurant         Clothing Store   
9               Aquarium                 Hotel                   Café   
10                 Hotel                  Café             Restaurant   
11                  Café            Restaurant                  Hotel   
12            Restaurant                   Bar    Japanese Restaurant   
13                  Café    Mexican Restaurant  Vietnamese Restaurant   
15                  Café              Beer Bar     Seafood Restaurant   
16                  Café                 Hotel                    Gym   

            5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1                          Market    Italian Restaurant                  Café   
2                      Restaurant                   Pub           Men's Store   
4             Japanese Restaurant        Cosmetics Shop    Italian Restaurant   
5                      Restaurant   American Restaurant          Cocktail Bar   
6                      Restaurant    Seafood Restaurant              Beer Bar   
7                Department Store           Salad Place          Burger Joint   
8                           Hotel                   Gym       Thai Restaurant   
9                      Restaurant               Brewery    Italian Restaurant   
10             Seafood Restaurant           Salad Place   Japanese Restaurant   
11            American Restaurant                   Gym    Seafood Restaurant   
12                      Bookstore        Sandwich Place                Bakery   
13  Vegetarian / Vegan Restaurant           Pizza Place  Caribbean Restaurant   
15                     Restaurant   Japanese Restaurant    Italian Restaurant   
16            Japanese Restaurant            Restaurant      Asian Restaurant   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1                 Restaurant   American Restaurant               Pharmacy  
2   Mediterranean Restaurant                 Hotel            Yoga Studio  
4                      Plaza           Pizza Place              Bookstore  
5             Lingerie Store        Farmers Market                    Gym  
6               Cocktail Bar                  Café       Greek Restaurant  
7            Bubble Tea Shop   Japanese Restaurant            Yoga Studio  
8                        Bar            Steakhouse       Sushi Restaurant  
9             Scenic Lookout   Fried Chicken Joint                  Plaza  
10        Italian Restaurant   American Restaurant       Sushi Restaurant  
11       J

In [46]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 3, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto               3       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14        Airport Lounge       Harbor / Marina                 Plane   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14         Boat or Ferry   Rental Car Location              Boutique   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14      Sculpture Garden               Airport       Airport Terminal

In [47]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 4, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
17  Downtown Toronto               4         Grocery Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17                  Café                  Park             Nightclub   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17                 Diner            Restaurant    Italian Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17            Baby Store           Candy Store     Athletics & Sports